File: src/domain/timer/TimerDomainEngine.ts
import {
  TimerSnapshot,
  EngineConfig,
  TimerEvent,
  TransitionResult,
  TimerEffect,
} from './types';

export class TimerDomainEngine {
  public static processEvent(
    snapshot: TimerSnapshot,
    event: TimerEvent,
    config: EngineConfig,
    now: number = event.timestamp
  ): TransitionResult {
    switch (event.type) {
      case 'START':
        return this.handleStart(snapshot, config, now);
      case 'STOP':
        return this.handleStop(snapshot, event);
      case 'TIMER_TRIGGER':
        return this.handleTimerTrigger(snapshot, event, config, now);
      case 'ACKNOWLEDGE':
        return this.handleAcknowledge(snapshot, event, config, now);
      case 'DRINK':
        return this.handleDrink(snapshot, event, config, now);
      case 'ALERT_TIMEOUT':
        return this.handleAlertTimeout(snapshot, event, config, now);
      case 'RECOVER':
        return this.handleRecover(snapshot, config, now);
    }
  }

  private static handleStart(
    snapshot: TimerSnapshot,
    config: EngineConfig,
    now: number
  ): TransitionResult {
    const nextGen = snapshot.timerGeneration + 1;
    const nextTriggerAt = calculateNextTriggerAt(now, config);
    const newSnapshot: TimerSnapshot = {
      state: 'WAITING',
      timerGeneration: nextGen,
      alertId: null,
      lastAlertId: snapshot.lastAlertId,
      nextTriggerAt,
      alertStartedAt: null,
      alertDeadlineAt: null,
    };

    const effects: TimerEffect[] = [];
    if (snapshot.timerGeneration > 0) {
      effects.push({ type: 'CancelTimer', timerGeneration: snapshot.timerGeneration });
      effects.push({ type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: snapshot.alertId });
      effects.push({ type: 'CancelReminder' });
    }
    effects.push({ type: 'PersistState', snapshot: newSnapshot });
    effects.push({ type: 'ScheduleTimer', triggerAtMillis: nextTriggerAt, timerGeneration: nextGen });

    return { newSnapshot, effects };
  }

  private static handleStop(
    snapshot: TimerSnapshot,
    event: TimerEvent
  ): TransitionResult {
    if (event.timerGeneration === undefined || event.timerGeneration === null || event.timerGeneration !== snapshot.timerGeneration) {
      return { newSnapshot: snapshot, effects: [] };
    }

    const newSnapshot: TimerSnapshot = {
      ...snapshot,
      state: 'STOPPED',
      alertId: null,
      nextTriggerAt: null,
      alertStartedAt: null,
      alertDeadlineAt: null,
    };

    const effects: TimerEffect[] = [
      { type: 'PersistState', snapshot: newSnapshot },
      { type: 'CancelTimer', timerGeneration: snapshot.timerGeneration },
      { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: snapshot.alertId },
      { type: 'CancelReminder' },
    ];

    return { newSnapshot, effects };
  }

  private static handleTimerTrigger(
    snapshot: TimerSnapshot,
    event: TimerEvent,
    config: EngineConfig,
    now: number
  ): TransitionResult {
    if (snapshot.state !== 'WAITING') {
      return { newSnapshot: snapshot, effects: [] };
    }
    if (event.timerGeneration === undefined || event.timerGeneration === null || event.timerGeneration !== snapshot.timerGeneration) {
      return { newSnapshot: snapshot, effects: [] };
    }

    const inQuiet = config.quietHoursEnabled && isInQuietHoursAt(now, config.quietHoursStart, config.quietHoursEnd);
    const dayActive = isDayActiveAt(now, config.activeDays);

    if (inQuiet || !dayActive) {
      const nextTriggerAt = calculateNextTriggerAt(now, config);
      const newSnapshot: TimerSnapshot = {
        ...snapshot,
        nextTriggerAt,
      };
      const effects: TimerEffect[] = [
        { type: 'PersistState', snapshot: newSnapshot },
        { type: 'ScheduleTimer', triggerAtMillis: nextTriggerAt, timerGeneration: snapshot.timerGeneration },
      ];
      return { newSnapshot, effects };
    }

    const nextAlertId = Math.max(1000, snapshot.lastAlertId) + 1;
    const alertDurationMs = Math.max(5, config.alertDurationSeconds) * 1000;
    const alertStartedAt = now;
    const alertDeadlineAt = now + alertDurationMs;

    const newSnapshot: TimerSnapshot = {
      ...snapshot,
      state: 'ALERT_ACTIVE',
      alertId: nextAlertId,
      lastAlertId: nextAlertId,
      alertStartedAt,
      alertDeadlineAt,
      nextTriggerAt: null,
    };

    const effects: TimerEffect[] = [
      { type: 'PersistState', snapshot: newSnapshot },
      {
        type: 'ShowReminder',
        title: null,
        body: null,
        timerGeneration: snapshot.timerGeneration,
        alertId: nextAlertId,
      },
      {
        type: 'ScheduleAlertTimeout',
        deadlineAtMillis: alertDeadlineAt,
        timerGeneration: snapshot.timerGeneration,
        alertId: nextAlertId,
      },
    ];

    return { newSnapshot, effects };
  }

  private static handleAcknowledge(
    snapshot: TimerSnapshot,
    event: TimerEvent,
    config: EngineConfig,
    now: number
  ): TransitionResult {
    if (snapshot.state !== 'ALERT_ACTIVE') {
      return { newSnapshot: snapshot, effects: [] };
    }
    if (event.timerGeneration === undefined || event.timerGeneration === null || event.timerGeneration !== snapshot.timerGeneration) {
      return { newSnapshot: snapshot, effects: [] };
    }
    if (snapshot.alertId === null || event.alertId === undefined || event.alertId === null || event.alertId !== snapshot.alertId) {
      return { newSnapshot: snapshot, effects: [] };
    }

    const activeAlertId = snapshot.alertId;
    const isCountdown = config.mode === 'countdown';

    if (isCountdown) {
      const newSnapshot: TimerSnapshot = {
        ...snapshot,
        state: 'STOPPED',
        alertId: null,
        nextTriggerAt: null,
        alertStartedAt: null,
        alertDeadlineAt: null,
      };
      const effects: TimerEffect[] = [
        { type: 'PersistState', snapshot: newSnapshot },
        { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: activeAlertId },
        { type: 'CancelReminder' },
        { type: 'RecordAck' },
      ];
      return { newSnapshot, effects };
    } else {
      const nextTriggerAt = calculateNextTriggerAt(now, config);
      const newSnapshot: TimerSnapshot = {
        ...snapshot,
        state: 'WAITING',
        alertId: null,
        nextTriggerAt,
        alertStartedAt: null,
        alertDeadlineAt: null,
      };
      const effects: TimerEffect[] = [
        { type: 'PersistState', snapshot: newSnapshot },
        { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: activeAlertId },
        { type: 'CancelReminder' },
        { type: 'RecordAck' },
        { type: 'ScheduleTimer', triggerAtMillis: nextTriggerAt, timerGeneration: snapshot.timerGeneration },
      ];
      return { newSnapshot, effects };
    }
  }

  private static handleDrink(
    snapshot: TimerSnapshot,
    event: TimerEvent,
    config: EngineConfig,
    now: number
  ): TransitionResult {
    if (snapshot.state !== 'ALERT_ACTIVE') {
      return { newSnapshot: snapshot, effects: [] };
    }
    if (event.timerGeneration === undefined || event.timerGeneration === null || event.timerGeneration !== snapshot.timerGeneration) {
      return { newSnapshot: snapshot, effects: [] };
    }
    if (snapshot.alertId === null || event.alertId === undefined || event.alertId === null || event.alertId !== snapshot.alertId) {
      return { newSnapshot: snapshot, effects: [] };
    }

    const activeAlertId = snapshot.alertId;
    const intake = Math.max(0, config.intakePerAlertMl);
    const isCountdown = config.mode === 'countdown';

    if (isCountdown) {
      const newSnapshot: TimerSnapshot = {
        ...snapshot,
        state: 'STOPPED',
        alertId: null,
        nextTriggerAt: null,
        alertStartedAt: null,
        alertDeadlineAt: null,
      };
      const effects: TimerEffect[] = [
        { type: 'PersistState', snapshot: newSnapshot },
        { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: activeAlertId },
        { type: 'CancelReminder' },
        { type: 'RecordDrink', amountMl: intake },
        { type: 'RecordAck' },
      ];
      return { newSnapshot, effects };
    } else {
      const nextTriggerAt = calculateNextTriggerAt(now, config);
      const newSnapshot: TimerSnapshot = {
        ...snapshot,
        state: 'WAITING',
        alertId: null,
        nextTriggerAt,
        alertStartedAt: null,
        alertDeadlineAt: null,
      };
      const effects: TimerEffect[] = [
        { type: 'PersistState', snapshot: newSnapshot },
        { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: activeAlertId },
        { type: 'CancelReminder' },
        { type: 'RecordDrink', amountMl: intake },
        { type: 'RecordAck' },
        { type: 'ScheduleTimer', triggerAtMillis: nextTriggerAt, timerGeneration: snapshot.timerGeneration },
      ];
      return { newSnapshot, effects };
    }
  }

  private static handleAlertTimeout(
    snapshot: TimerSnapshot,
    event: TimerEvent,
    config: EngineConfig,
    now: number
  ): TransitionResult {
    if (snapshot.state !== 'ALERT_ACTIVE') {
      return { newSnapshot: snapshot, effects: [] };
    }
    if (event.timerGeneration === undefined || event.timerGeneration === null || event.timerGeneration !== snapshot.timerGeneration) {
      return { newSnapshot: snapshot, effects: [] };
    }
    if (snapshot.alertId === null || event.alertId === undefined || event.alertId === null || event.alertId !== snapshot.alertId) {
      return { newSnapshot: snapshot, effects: [] };
    }

    const activeAlertId = snapshot.alertId;
    const durationMs =
      snapshot.alertDeadlineAt !== null && snapshot.alertStartedAt !== null
        ? snapshot.alertDeadlineAt - snapshot.alertStartedAt
        : Math.max(5, config.alertDurationSeconds) * 1000;
    const durationSec = Math.max(5, Math.floor(durationMs / 1000));
    const isCountdown = config.mode === 'countdown';

    if (isCountdown) {
      const newSnapshot: TimerSnapshot = {
        ...snapshot,
        state: 'STOPPED',
        alertId: null,
        nextTriggerAt: null,
        alertStartedAt: null,
        alertDeadlineAt: null,
      };
      const effects: TimerEffect[] = [
        { type: 'PersistState', snapshot: newSnapshot },
        { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: activeAlertId },
        { type: 'CancelReminder' },
        { type: 'RecordMissed' },
        { type: 'ShowMissedNotification', alertDurationSeconds: durationSec },
      ];
      return { newSnapshot, effects };
    } else {
      const nextTriggerAt = calculateNextTriggerAt(now, config);
      const newSnapshot: TimerSnapshot = {
        ...snapshot,
        state: 'WAITING',
        alertId: null,
        nextTriggerAt,
        alertStartedAt: null,
        alertDeadlineAt: null,
      };
      const effects: TimerEffect[] = [
        { type: 'PersistState', snapshot: newSnapshot },
        { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: activeAlertId },
        { type: 'CancelReminder' },
        { type: 'RecordMissed' },
        { type: 'ShowMissedNotification', alertDurationSeconds: durationSec },
        { type: 'ScheduleTimer', triggerAtMillis: nextTriggerAt, timerGeneration: snapshot.timerGeneration },
      ];
      return { newSnapshot, effects };
    }
  }

  private static handleRecover(
    snapshot: TimerSnapshot,
    config: EngineConfig,
    now: number
  ): TransitionResult {
    switch (snapshot.state) {
      case 'STOPPED': {
        const effects: TimerEffect[] = [
          { type: 'PersistState', snapshot },
          { type: 'CancelTimer', timerGeneration: snapshot.timerGeneration },
          { type: 'CancelAlertTimeout', timerGeneration: snapshot.timerGeneration, alertId: snapshot.alertId },
        ];
        return { newSnapshot: snapshot, effects };
      }
      case 'WAITING': {
        const targetAt = snapshot.nextTriggerAt ?? now;
        if (now >= targetAt) {
          const overdueMs = now - targetAt;
          const maxGraceMs = Math.max(5, config.alertDurationSeconds) * 1000 * 2;
          if (overdueMs <= maxGraceMs) {
            const triggerEvent: TimerEvent = {
              type: 'TIMER_TRIGGER',
              timerGeneration: snapshot.timerGeneration,
              timestamp: now,
            };
            return this.processEvent(snapshot, triggerEvent, config, now);
          } else {
            const nextTriggerAt = calculateNextTriggerAt(now, config);
            const newSnapshot: TimerSnapshot = {
              ...snapshot,
              nextTriggerAt,
            };
            const effects: TimerEffect[] = [
              { type: 'PersistState', snapshot: newSnapshot },
              { type: 'ScheduleTimer', triggerAtMillis: nextTriggerAt, timerGeneration: snapshot.timerGeneration },
            ];
            return { newSnapshot, effects };
          }
        } else {
          const effects: TimerEffect[] = [
            { type: 'PersistState', snapshot },
            { type: 'ScheduleTimer', triggerAtMillis: targetAt, timerGeneration: snapshot.timerGeneration },
          ];
          return { newSnapshot: snapshot, effects };
        }
      }
      case 'ALERT_ACTIVE': {
        const deadlineAt = snapshot.alertDeadlineAt ?? now;
        if (now >= deadlineAt) {
          const timeoutEvent: TimerEvent = {
            type: 'ALERT_TIMEOUT',
            timerGeneration: snapshot.timerGeneration,
            alertId: snapshot.alertId,
            timestamp: now,
          };
          return this.processEvent(snapshot, timeoutEvent, config, now);
        } else {
          const activeAlertId = snapshot.alertId ?? Math.max(1001, snapshot.lastAlertId);
          const effects: TimerEffect[] = [
            { type: 'PersistState', snapshot },
            {
              type: 'ShowReminder',
              title: null,
              body: null,
              timerGeneration: snapshot.timerGeneration,
              alertId: activeAlertId,
            },
            {
              type: 'ScheduleAlertTimeout',
              deadlineAtMillis: deadlineAt,
              timerGeneration: snapshot.timerGeneration,
              alertId: activeAlertId,
            },
          ];
          return { newSnapshot: snapshot, effects };
        }
      }
    }
  }
}

export function calculateNextTriggerAt(now: number, config: EngineConfig): number {
  if (config.mode === 'countdown') {
    const totalMs = (config.countdownMinutes * 60 + config.countdownSeconds) * 1000;
    return now + Math.max(1000, totalMs);
  }

  let candidate: number;
  if (config.intervalMode !== 'clock') {
    candidate = now + Math.max(1, config.intervalMinutes) * 60_000;
  } else {
    const step = Math.min(90, Math.max(30, config.clockIntervalMinutes));
    const d = new Date(now);
    const currentMinute = d.getHours() * 60 + d.getMinutes();
    const nextSlot = (Math.floor(currentMinute / step) + 1) * step;
    const targetHour = Math.floor(nextSlot / 60) % 24;
    const targetMin = nextSlot % 60;

    const res = new Date(now);
    res.setHours(targetHour, targetMin, 0, 0);
    if (nextSlot >= 24 * 60 || res.getTime() <= now) {
      res.setDate(res.getDate() + 1);
    }
    candidate = res.getTime();
  }

  return adjustForQuietHoursAndActiveDays(candidate, config);
}

export function adjustForQuietHoursAndActiveDays(candidateTrigger: number, config: EngineConfig): number {
  let time = candidateTrigger;
  let iterations = 0;
  while (iterations < 14 * 24) {
    const d = new Date(time);
    const dayIndex = d.getDay(); // 0 = Sunday, 1 = Monday, ..., 6 = Saturday
    const isDayActive = config.activeDays && config.activeDays[dayIndex] !== undefined ? config.activeDays[dayIndex] : true;

    if (!isDayActive) {
      d.setDate(d.getDate() + 1);
      d.setHours(0, 0, 0, 0);
      time = d.getTime();
      iterations++;
      continue;
    }

    if (config.quietHoursEnabled && isInQuietHoursAt(time, config.quietHoursStart, config.quietHoursEnd)) {
      const endParsed = parseTimeOfDay(config.quietHoursEnd);
      if (endParsed) {
        d.setHours(endParsed.h, endParsed.m, 0, 0);
        if (d.getTime() <= time) {
          d.setDate(d.getDate() + 1);
        }
        time = d.getTime();
        iterations++;
        continue;
      }
    }

    break;
  }
  return time;
}

export function isInQuietHoursAt(timeMillis: number, startStr: string, endStr: string): boolean {
  const d = new Date(timeMillis);
  const nowMins = d.getHours() * 60 + d.getMinutes();
  const startParsed = parseTimeOfDay(startStr);
  const endParsed = parseTimeOfDay(endStr);
  const s = startParsed ? startParsed.h * 60 + startParsed.m : 23 * 60;
  const e = endParsed ? endParsed.h * 60 + endParsed.m : 7 * 60;
  if (s < e) {
    return nowMins >= s && nowMins < e;
  } else {
    return nowMins >= s || nowMins < e;
  }
}

export function isDayActiveAt(timeMillis: number, activeDays: boolean[]): boolean {
  const d = new Date(timeMillis);
  const dayIndex = d.getDay(); // 0 = Sunday..6 = Saturday
  return activeDays && activeDays[dayIndex] !== undefined ? activeDays[dayIndex] : true;
}

export function parseTimeOfDay(timeStr: string): { h: number; m: number } | null {
  if (!timeStr) return null;
  const parts = timeStr.trim().split(':');
  if (parts.length === 2) {
    const h = parseInt(parts[0], 10);
    const m = parseInt(parts[1], 10);
    if (!isNaN(h) && !isNaN(m) && h >= 0 && h <= 23 && m >= 0 && m <= 59) {
      return { h, m };
    }
  }
  return null;
}

File: android/app/src/main/java/com/hydrotimer/app/TimerDomainEngine.kt
package com.hydrotimer.app

import java.util.Calendar

data class TransitionResult(
    val newSnapshot: TimerSnapshot,
    val effects: List<TimerEffect>
)

object TimerDomainEngine {

    fun processEvent(
        snapshot: TimerSnapshot,
        event: TimerEvent,
        config: EngineConfig,
        now: Long = event.timestamp
    ): TransitionResult {
        return when (event.type) {
            TimerEventType.START -> handleStart(snapshot, config, now)
            TimerEventType.STOP -> handleStop(snapshot, event)
            TimerEventType.TIMER_TRIGGER -> handleTimerTrigger(snapshot, event, config, now)
            TimerEventType.ACKNOWLEDGE -> handleAcknowledge(snapshot, event, config, now)
            TimerEventType.DRINK -> handleDrink(snapshot, event, config, now)
            TimerEventType.ALERT_TIMEOUT -> handleAlertTimeout(snapshot, event, config, now)
            TimerEventType.RECOVER -> handleRecover(snapshot, config, now)
        }
    }

    private fun handleStart(
        snapshot: TimerSnapshot,
        config: EngineConfig,
        now: Long
    ): TransitionResult {
        val nextGen = snapshot.timerGeneration + 1L
        val nextTriggerAt = calculateNextTriggerAt(now, config)
        val newSnapshot = TimerSnapshot(
            state = TimerState.WAITING,
            timerGeneration = nextGen,
            alertId = null,
            lastAlertId = snapshot.lastAlertId,
            nextTriggerAt = nextTriggerAt,
            alertStartedAt = null,
            alertDeadlineAt = null
        )

        val effects = mutableListOf<TimerEffect>()
        if (snapshot.timerGeneration > 0L) {
            effects.add(TimerEffect.CancelTimer(snapshot.timerGeneration))
            effects.add(TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, snapshot.alertId))
            effects.add(TimerEffect.CancelReminder)
        }
        effects.add(TimerEffect.ScheduleTimer(nextTriggerAt, nextGen))
        effects.add(TimerEffect.PersistState(newSnapshot))

        return TransitionResult(newSnapshot, effects)
    }

    private fun handleStop(
        snapshot: TimerSnapshot,
        event: TimerEvent
    ): TransitionResult {
        if (event.timerGeneration != null && event.timerGeneration != snapshot.timerGeneration) {
            return TransitionResult(snapshot, emptyList())
        }

        val newSnapshot = snapshot.copy(
            state = TimerState.STOPPED,
            alertId = null,
            nextTriggerAt = null,
            alertStartedAt = null,
            alertDeadlineAt = null
        )

        val effects = listOf(
            TimerEffect.CancelTimer(snapshot.timerGeneration),
            TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, snapshot.alertId),
            TimerEffect.CancelReminder,
            TimerEffect.PersistState(newSnapshot)
        )

        return TransitionResult(newSnapshot, effects)
    }

    private fun handleTimerTrigger(
        snapshot: TimerSnapshot,
        event: TimerEvent,
        config: EngineConfig,
        now: Long
    ): TransitionResult {
        if (snapshot.state != TimerState.WAITING) {
            return TransitionResult(snapshot, emptyList())
        }
        if (event.timerGeneration != null && event.timerGeneration != snapshot.timerGeneration) {
            return TransitionResult(snapshot, emptyList())
        }

        val inQuiet = config.quietHoursEnabled && isInQuietHoursAt(now, config.quietHoursStart, config.quietHoursEnd)
        val dayActive = isDayActiveAt(now, config.activeDays)

        if (inQuiet || !dayActive) {
            val nextTriggerAt = calculateNextTriggerAt(now, config)
            val newSnapshot = snapshot.copy(nextTriggerAt = nextTriggerAt)
            val effects = listOf(
                TimerEffect.ScheduleTimer(nextTriggerAt, snapshot.timerGeneration),
                TimerEffect.PersistState(newSnapshot)
            )
            return TransitionResult(newSnapshot, effects)
        }

        val nextAlertId = Math.max(1000L, snapshot.lastAlertId) + 1L
        val alertDurationMs = Math.max(5, config.alertDurationSeconds) * 1000L
        val alertStartedAt = now
        val alertDeadlineAt = now + alertDurationMs

        val newSnapshot = snapshot.copy(
            state = TimerState.ALERT_ACTIVE,
            alertId = nextAlertId,
            lastAlertId = nextAlertId,
            alertStartedAt = alertStartedAt,
            alertDeadlineAt = alertDeadlineAt,
            nextTriggerAt = null
        )

        val effects = listOf(
            TimerEffect.ShowReminder(title = null, body = null, timerGeneration = snapshot.timerGeneration, alertId = nextAlertId),
            TimerEffect.ScheduleAlertTimeout(deadlineAtMillis = alertDeadlineAt, timerGeneration = snapshot.timerGeneration, alertId = nextAlertId),
            TimerEffect.PersistState(newSnapshot)
        )

        return TransitionResult(newSnapshot, effects)
    }

    private fun handleAcknowledge(
        snapshot: TimerSnapshot,
        event: TimerEvent,
        config: EngineConfig,
        now: Long
    ): TransitionResult {
        if (snapshot.state != TimerState.ALERT_ACTIVE) {
            return TransitionResult(snapshot, emptyList())
        }
        if (event.timerGeneration != null && event.timerGeneration != snapshot.timerGeneration) {
            return TransitionResult(snapshot, emptyList())
        }
        if (snapshot.alertId == null || (event.alertId != null && event.alertId != snapshot.alertId)) {
            return TransitionResult(snapshot, emptyList())
        }

        val activeAlertId = snapshot.alertId
        val isCountdown = config.mode == "countdown"

        if (isCountdown) {
            val newSnapshot = snapshot.copy(
                state = TimerState.STOPPED,
                alertId = null,
                nextTriggerAt = null,
                alertStartedAt = null,
                alertDeadlineAt = null
            )
            val effects = listOf(
                TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, activeAlertId),
                TimerEffect.CancelReminder,
                TimerEffect.RecordAck,
                TimerEffect.PersistState(newSnapshot)
            )
            return TransitionResult(newSnapshot, effects)
        } else {
            val nextTriggerAt = calculateNextTriggerAt(now, config)
            val newSnapshot = snapshot.copy(
                state = TimerState.WAITING,
                alertId = null,
                nextTriggerAt = nextTriggerAt,
                alertStartedAt = null,
                alertDeadlineAt = null
            )
            val effects = listOf(
                TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, activeAlertId),
                TimerEffect.CancelReminder,
                TimerEffect.RecordAck,
                TimerEffect.ScheduleTimer(nextTriggerAt, snapshot.timerGeneration),
                TimerEffect.PersistState(newSnapshot)
            )
            return TransitionResult(newSnapshot, effects)
        }
    }

    private fun handleDrink(
        snapshot: TimerSnapshot,
        event: TimerEvent,
        config: EngineConfig,
        now: Long
    ): TransitionResult {
        if (snapshot.state != TimerState.ALERT_ACTIVE) {
            return TransitionResult(snapshot, emptyList())
        }
        if (event.timerGeneration != null && event.timerGeneration != snapshot.timerGeneration) {
            return TransitionResult(snapshot, emptyList())
        }
        if (snapshot.alertId == null || (event.alertId != null && event.alertId != snapshot.alertId)) {
            return TransitionResult(snapshot, emptyList())
        }

        val activeAlertId = snapshot.alertId
        val intake = Math.max(0, config.intakePerAlertMl)
        val isCountdown = config.mode == "countdown"

        if (isCountdown) {
            val newSnapshot = snapshot.copy(
                state = TimerState.STOPPED,
                alertId = null,
                nextTriggerAt = null,
                alertStartedAt = null,
                alertDeadlineAt = null
            )
            val effects = listOf(
                TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, activeAlertId),
                TimerEffect.CancelReminder,
                TimerEffect.RecordDrink(intake),
                TimerEffect.RecordAck,
                TimerEffect.PersistState(newSnapshot)
            )
            return TransitionResult(newSnapshot, effects)
        } else {
            val nextTriggerAt = calculateNextTriggerAt(now, config)
            val newSnapshot = snapshot.copy(
                state = TimerState.WAITING,
                alertId = null,
                nextTriggerAt = nextTriggerAt,
                alertStartedAt = null,
                alertDeadlineAt = null
            )
            val effects = listOf(
                TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, activeAlertId),
                TimerEffect.CancelReminder,
                TimerEffect.RecordDrink(intake),
                TimerEffect.RecordAck,
                TimerEffect.ScheduleTimer(nextTriggerAt, snapshot.timerGeneration),
                TimerEffect.PersistState(newSnapshot)
            )
            return TransitionResult(newSnapshot, effects)
        }
    }

    private fun handleAlertTimeout(
        snapshot: TimerSnapshot,
        event: TimerEvent,
        config: EngineConfig,
        now: Long
    ): TransitionResult {
        if (snapshot.state != TimerState.ALERT_ACTIVE) {
            return TransitionResult(snapshot, emptyList())
        }
        if (event.timerGeneration != null && event.timerGeneration != snapshot.timerGeneration) {
            return TransitionResult(snapshot, emptyList())
        }
        if (snapshot.alertId == null || (event.alertId != null && event.alertId != snapshot.alertId)) {
            return TransitionResult(snapshot, emptyList())
        }

        val activeAlertId = snapshot.alertId
        val durationSec = Math.max(5, config.alertDurationSeconds)
        val isCountdown = config.mode == "countdown"

        if (isCountdown) {
            val newSnapshot = snapshot.copy(
                state = TimerState.STOPPED,
                alertId = null,
                nextTriggerAt = null,
                alertStartedAt = null,
                alertDeadlineAt = null
            )
            val effects = listOf(
                TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, activeAlertId),
                TimerEffect.CancelReminder,
                TimerEffect.RecordMissed,
                TimerEffect.ShowMissedNotification(durationSec),
                TimerEffect.PersistState(newSnapshot)
            )
            return TransitionResult(newSnapshot, effects)
        } else {
            val nextTriggerAt = calculateNextTriggerAt(now, config)
            val newSnapshot = snapshot.copy(
                state = TimerState.WAITING,
                alertId = null,
                nextTriggerAt = nextTriggerAt,
                alertStartedAt = null,
                alertDeadlineAt = null
            )
            val effects = listOf(
                TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, activeAlertId),
                TimerEffect.CancelReminder,
                TimerEffect.RecordMissed,
                TimerEffect.ShowMissedNotification(durationSec),
                TimerEffect.ScheduleTimer(nextTriggerAt, snapshot.timerGeneration),
                TimerEffect.PersistState(newSnapshot)
            )
            return TransitionResult(newSnapshot, effects)
        }
    }

    private fun handleRecover(
        snapshot: TimerSnapshot,
        config: EngineConfig,
        now: Long
    ): TransitionResult {
        return when (snapshot.state) {
            TimerState.STOPPED -> {
                val effects = listOf(
                    TimerEffect.CancelTimer(snapshot.timerGeneration),
                    TimerEffect.CancelAlertTimeout(snapshot.timerGeneration, snapshot.alertId),
                    TimerEffect.PersistState(snapshot)
                )
                TransitionResult(snapshot, effects)
            }
            TimerState.WAITING -> {
                val targetAt = snapshot.nextTriggerAt ?: now
                if (now >= targetAt) {
                    val triggerEvent = TimerEvent(
                        type = TimerEventType.TIMER_TRIGGER,
                        timerGeneration = snapshot.timerGeneration,
                        timestamp = now
                    )
                    processEvent(snapshot, triggerEvent, config, now)
                } else {
                    val effects = listOf(
                        TimerEffect.ScheduleTimer(targetAt, snapshot.timerGeneration),
                        TimerEffect.PersistState(snapshot)
                    )
                    TransitionResult(snapshot, effects)
                }
            }
            TimerState.ALERT_ACTIVE -> {
                val deadlineAt = snapshot.alertDeadlineAt ?: now
                if (now >= deadlineAt) {
                    val timeoutEvent = TimerEvent(
                        type = TimerEventType.ALERT_TIMEOUT,
                        timerGeneration = snapshot.timerGeneration,
                        alertId = snapshot.alertId,
                        timestamp = now
                    )
                    processEvent(snapshot, timeoutEvent, config, now)
                } else {
                    val activeAlertId = snapshot.alertId ?: Math.max(1001L, snapshot.lastAlertId)
                    val effects = listOf(
                        TimerEffect.ShowReminder(title = null, body = null, timerGeneration = snapshot.timerGeneration, alertId = activeAlertId),
                        TimerEffect.ScheduleAlertTimeout(deadlineAtMillis = deadlineAt, timerGeneration = snapshot.timerGeneration, alertId = activeAlertId),
                        TimerEffect.PersistState(snapshot)
                    )
                    TransitionResult(snapshot, effects)
                }
            }
        }
    }

    fun calculateNextTriggerAt(now: Long, config: EngineConfig): Long {
        if (config.mode == "countdown") {
            val totalMs = (config.countdownMinutes * 60 + config.countdownSeconds) * 1000L
            return now + Math.max(1000L, totalMs)
        }

        var candidate: Long
        if (config.intervalMode != "clock") {
            candidate = now + Math.max(1, config.intervalMinutes) * 60_000L
        } else {
            val step = config.clockIntervalMinutes.coerceIn(30, 90)
            val cal = Calendar.getInstance().apply { timeInMillis = now }
            val currentMinute = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE)
            val nextSlot = ((currentMinute / step) + 1) * step
            val res = Calendar.getInstance().apply {
                timeInMillis = now
                set(Calendar.HOUR_OF_DAY, (nextSlot / 60) % 24)
                set(Calendar.MINUTE, nextSlot % 60)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
                if (nextSlot >= 24 * 60 || timeInMillis <= now) {
                    add(Calendar.DAY_OF_YEAR, 1)
                }
            }
            candidate = res.timeInMillis
        }

        return adjustForQuietHoursAndActiveDays(candidate, config)
    }

    fun adjustForQuietHoursAndActiveDays(candidateTrigger: Long, config: EngineConfig): Long {
        var time = candidateTrigger
        var iterations = 0
        while (iterations < 14 * 24) {
            val cal = Calendar.getInstance().apply { timeInMillis = time }
            val dayOfWeek = cal.get(Calendar.DAY_OF_WEEK) // 1=Sun..7=Sat
            val dayIndex = dayOfWeek - 1
            val isDayActive = config.activeDays.getOrElse(dayIndex) { true }

            if (!isDayActive) {
                cal.add(Calendar.DAY_OF_YEAR, 1)
                cal.set(Calendar.HOUR_OF_DAY, 0)
                cal.set(Calendar.MINUTE, 0)
                cal.set(Calendar.SECOND, 0)
                cal.set(Calendar.MILLISECOND, 0)
                time = cal.timeInMillis
                iterations++
                continue
            }

            if (config.quietHoursEnabled && isInQuietHoursAt(time, config.quietHoursStart, config.quietHoursEnd)) {
                val endParsed = parseTimeOfDay(config.quietHoursEnd)
                if (endParsed != null) {
                    val (qEndH, qEndM) = endParsed
                    cal.set(Calendar.HOUR_OF_DAY, qEndH)
                    cal.set(Calendar.MINUTE, qEndM)
                    cal.set(Calendar.SECOND, 0)
                    cal.set(Calendar.MILLISECOND, 0)
                    if (cal.timeInMillis <= time) {
                        cal.add(Calendar.DAY_OF_YEAR, 1)
                    }
                    time = cal.timeInMillis
                    iterations++
                    continue
                }
            }

            break
        }
        return time
    }

    fun isInQuietHoursAt(timeMillis: Long, startStr: String, endStr: String): Boolean {
        val cal = Calendar.getInstance().apply { timeInMillis = timeMillis }
        val nowMins = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE)
        val s = parseTimeOfDay(startStr)?.let { it.first * 60 + it.second } ?: (23 * 60)
        val e = parseTimeOfDay(endStr)?.let { it.first * 60 + it.second } ?: (7 * 60)
        return if (s < e) {
            nowMins in s until e
        } else {
            nowMins >= s || nowMins < e
        }
    }

    fun isDayActiveAt(timeMillis: Long, activeDays: BooleanArray): Boolean {
        val cal = Calendar.getInstance().apply { timeInMillis = timeMillis }
        val dayOfWeek = cal.get(Calendar.DAY_OF_WEEK) // 1=Sun..7=Sat
        val dayIndex = dayOfWeek - 1
        return activeDays.getOrElse(dayIndex) { true }
    }

    private fun parseTimeOfDay(timeStr: String): Pair<Int, Int>? {
        val parts = timeStr.trim().split(":")
        if (parts.size == 2) {
            val h = parts[0].toIntOrNull()
            val m = parts[1].toIntOrNull()
            if (h != null && m != null && h in 0..23 && m in 0..59) {
                return Pair(h, m)
            }
        }
        return null
    }
}
